db/cluster/cluster.go

```diff
@@ -0,0 +1,372 @@
+const defaultUpdateTopology = 30 * time.Second // 30 * time.Second
+
+const maxTimeLeaving = 20 * time.Second
+const failDuration = 60 * time.Second
+
+// RDSReadCluster provides db instances on demand. In comparison to default sql.DB with go-sql-driver/mysql it load balances across all read replicas.
+type RDSReadCluster interface {
+
+	// QueryContext executes a query that returns rows, typically a SELECT.
+	// The args are for any placeholder parameters in the query.
+	// Does automatic load balancing and retries on broken connection.
+	QueryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error)
+
+	// Query executes a query that returns rows, typically a SELECT.
+	// The args are for any placeholder parameters in the query.
+	Query(query string, args ...interface{}) (*sql.Rows, error)
+
+	// Close frees all resources. Make sure to finish all queries before calling Close.
+	Close() error
+}
+
+// rdsReadCluster is an implementaion of RDSReadCluster
+type rdsReadCluster struct {
+	opts Opts
+
+	// topology is the list of currently available server ids
+	topology   *topology
+	topologyMu sync.Mutex
+
+	topologyUpdates *ticker
+
+	// dbs is a map of db instances connecting to specific server
+	// map[server_id]*sql.DB
+	dbs   map[string]*sql.DB
+	dbsMu sync.Mutex
+}
+
+// New creates a new rdsReadCluster
+func New(opts Opts) *rdsReadCluster {
+	if opts.Port == 0 {
+		opts.Port = defaultPort
+	}
+	if opts.UpdateTopologyEvery == 0 {
+		opts.UpdateTopologyEvery = defaultUpdateTopology
+	}
+	if opts.User == "" || opts.Pass == "" || opts.Port == 0 || opts.Database == "" || opts.ReadEndpointURL == "" || opts.ClusterURLSuffix == "" || opts.UpdateTopologyEvery == 0 || opts.MaxConnectionsPerServer == 0 || opts.Log == nil {
+		panic("provide all options")
+	}
+	s := &rdsReadCluster{}
+	s.opts = opts
+
+	s.dbs = map[string]*sql.DB{}
+
+	s.topology = newTopology(topologyOpts{
+		MaxTimeLeaving: maxTimeLeaving,
+		OnLeave: func(id string) {
+			s.dbsMu.Lock()
+			defer s.dbsMu.Unlock()
+			if _, ok := s.dbs[id]; !ok {
+				// already left
+				return
+			}
+			go func(db *sql.DB) {
+				// do not block
+				db.Close()
+			}(s.dbs[id])
+			delete(s.dbs, id)
+		},
+		FailDuration: failDuration,
+		Log:          s.opts.Log,
+		Now:          time.Now,
+	})
+
+	err := s.updateTopologyInitial()
+	if err != nil {
+		panic(err)
+	}
+
+	s.setupTopologyTicker()
+	return s
+}
+
+var ErrCantConnect = errors.New("cluster: can't connect to any read replica")
+
+func (s *rdsReadCluster) QueryContext(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error) {
+
+	maxRetries := 10
+	for i := 0; i < maxRetries; i++ {
+		if i > 0 {
+			s.opts.Log("query retry", i)
+		}
+
+		db, host, err := s.loadBalancedDB()
+		if err != nil {
+			return nil, err
+		}
+
+		rows, err := db.QueryContext(ctx, query, args...)
+		if err != nil {
+			if i == 0 {
+				// retry once without marking the server as failed
+				continue
```

Comment:
ErrNoRows is only returned by Scan when QueryRow doesn't return a row. So it's not applicable in this case.

I've added a test case that checks that empty result set does not lead to error just in case.

But this does remind me, that I should handle an error with invalid query earlier since that does not require retry.